core/dbt/contracts/graph/nodes.py

```diff
@@ -539,29 +539,61 @@ def build_contract_checksum(self):
             self.contract.checksum = hashlib.new("sha256", data).hexdigest()
 
     def same_contract(self, old) -> bool:
+        # If the contract wasn't previously enforced
         if old.contract.enforced is False and self.contract.enforced is False:
             # Not a change
             return True
         if old.contract.enforced is False and self.contract.enforced is True:
             # A change, but not a breaking change
             return False
 
-        breaking_change_reasons = []
+        # Otherwise: the contract was previously enforced
+        self.build_contract_checksum()
+
+        # If the checksums match up, the contract has not changed, so same_contract: True
+        if self.contract.checksum == old.contract.checksum:
+            return True
+
+        # The checksums don't match up, so there has been a change.
+        # We need to determine if it's a **breaking** change.
+        # These are the categories of breaking changes:
+        contract_enforced_disabled: bool = False
+        columns_removed: List[str] = []
+        column_type_changes: List[Tuple[str, str, str]] = []
+
         if old.contract.enforced is True and self.contract.enforced is False:
-            # Breaking change: throw an error
-            # Note: we don't have contract.checksum for current node, so build
+            # Breaking change: the contract was previously enforced, and it no longer is
+            # Note: we don't have contract.checksum for current node, so build it now
             self.build_contract_checksum()
```

Comment:
have we not already done this up [here](https://github.com/dbt-labs/dbt-core/pull/7333/files#diff-42cfc99b2fcd89ab75957b8999bee2b1af822f7d39953ca99f1f285c170557c1R551)?

Reply:
I'm pretty sure we can remove both calls to `build_contract_checksum()`. It's being called within [`NodePatchParser.parse_patch`](https://github.com/dbt-labs/dbt-core/blob/main/core/dbt/parser/schemas.py#L999)

Reply:
Oh, that's a change. Initially I only had it building when contract.enforced is True, and it's been changed so that it always builds.

Reply:
Actually no, it's "build_contract_checksum" that checks for contract enforced. So it is a no-op, but it probably shouldn't be, because Michelle wanted to also capture that there were additional changes to the columns. So we should remove the check for contract.enforced in build_contract_checksum and only call it in parse_patch if contract.enforced is True, and THEN it won't be a no-op.

Reply:
Boy this logic makes your head hurt.

Reply:
@gshank Right 😅

We want the full column spec to be captured in the checksum, so that additions would cause a checksum mismatch — but if the checksums don't match, we then need to dig in and understand whether a column was actually removed/changed, or just added. I'm happy with the current implementation for now, but we can open up another ticket to revisit this logic.